src/execution_plan/execution_plan.c

```diff
@@ -8,695 +8,804 @@
+    OpBase *cartesianProduct = NULL;
+    if(connectedComponentsCount > 1) {
+        cartesianProduct = NewCartesianProductOp();
+        Vector_Push(ops, cartesianProduct);
     }
-}
 
-OpBase** ExecutionPlan_LocateOps(OpBase *root, OPType type) {
-    OpBase **ops = array_new(OpBase*, 0);
-    _ExecutionPlan_LocateOps(root, type, &ops);
-    return ops;
-}
+    // Keep track after all traversal operations along a pattern.
+    Vector *traversals = NewVector(OpBase*, 1);
+    OpBase *op;
 
-void ExecutionPlan_Taps(OpBase *root, OpBase ***taps) {
-    if(root == NULL) return;
-    if(root->type & OP_SCAN) *taps = array_append(*taps, root);
+    for(uint i = 0; i < connectedComponentsCount; i++) {
+        QueryGraph *cc = connectedComponents[i];
+        uint edge_count = array_len(cc->edges);
+        if(edge_count == 0) {
+            /* Node scan. */
+            QGNode *n = cc->nodes[0];
+            uint rec_idx = RecordMap_FindOrAddID(segment->record_map, n->id);
+            if(n->labelID != GRAPH_NO_LABEL) op = NewNodeByLabelScanOp(n, rec_idx);
+            else op = NewAllNodeScanOp(gc->g, n, rec_idx);
+            Vector_Push(traversals, op);
+            _UpdateResolvedVariables(resolved, op);
+        } else {
+            size_t expCount = 0;
+            AlgebraicExpression **exps = AlgebraicExpression_FromQueryGraph(cc, segment->record_map, &expCount);
+
+            // Reorder exps, to the most performant arrangement of evaluation.
+            orderExpressions(exps, expCount, segment->record_map, ft);
+
+            AlgebraicExpression *exp = exps[0];
+            selectEntryPoint(exp, segment->record_map, ft);
+
+            // Retrieve the AST ID for the source node
+            uint ast_id = exp->src_node->id;
+            // Convert to a Record ID
+            uint record_id = RecordMap_FindOrAddID(segment->record_map, ast_id);
+
+            // Create SCAN operation.
+            if(exp->src_node->label) {
+                /* Resolve source node by performing label scan,
+                 * in which case if the first algebraic expression operand
+                 * is a label matrix (diagonal) remove it, otherwise
+                 * the label matrix associated with source's label is located
+                 * within another traversal operation, for the timebeing do not
+                 * try to locate and remove it, there's no real harm except some performace hit
+                 * in keeping that label matrix. */
+                if(exp->operands[0].diagonal) AlgebraicExpression_RemoveTerm(exp, 0, NULL);
+                op = NewNodeByLabelScanOp(exp->src_node, record_id);
+            } else {
+                op = NewAllNodeScanOp(gc->g, exp->src_node, record_id);
+            }
 
-    for(int i = 0; i < root->childCount; i++) {
-        ExecutionPlan_Taps(root->children[i], taps);
-    }
-}
+            Vector_Push(traversals, op);
+            _UpdateResolvedVariables(resolved, op);
```

Comment:
Why do we constantly updating resolved vars, can't we do this at the very end by traversing the execution plan?